### ISSUE TITLE
Implement Zip file system support

### ISSUE THREAD

**Issue by WojciechMazur** [scala-native/scala-native#3453]
Native toolachain based on #3233 is currently not possible, becouse currently we're not able to operate on JARs containing NIR of dependencies. We should implement `java.nio.file.ZipFileSystemProvider` and it's other classes, which can provide support for jars. 
Apache Harmony does not implement this file system, so we'll need to reimplement it from scratch.

**Comment by armanbilge** [comment]
`ZipFileSystemProvider` was originally added to the JDK repository as a demo in https://github.com/openjdk/jdk/commit/2d526410d44047e2a5161f6b0df9175af45af013#diff-42f26f0fcad299efb87bb247990c0fb82b47dd0bf19874903179ef228a909be9.

This is notable because the original license is _not_ GPL, but as follows:
```
/*
 * Copyright (c) 2009, 2010, Oracle and/or its affiliates. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   - Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   - Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *   - Neither the name of Oracle nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
```

**Comment by armanbilge** [comment]
The commit that changed the license is https://github.com/openjdk/jdk/commit/0298d8c0fa27bdb01ff182faea083215abcfa176 when it became officially supported in the JDK. We should not look at that commit or any commit that followed.

The last commit with the liberally-licensed demo codes is here:
https://github.com/openjdk/jdk/tree/492411d1c6d8b34f3df00f152c9e4ad0ad5f97e3/jdk/src/share/demo/nio/zipfs

We can base our implementation on that.

**Comment by armanbilge** [comment]
For the record: the license for the demo appears to be the BSD-3 license.

https://opensource.org/license/bsd-3-clause/

**Comment by LeeTibbert** [comment]
@WojciechMazur, @ekrich, @armanbilge, at_Community

There seem to be some drive to get a native toolchain working.  

To that end, let me ask: Is anybody currently working on this issue?   If yes, godspeed.
If no, I can take a run at it.   

There are at least two issues where some direction/consensus would help, please advise:

1) A comment on the top of each zip file stating the commit SHA and that it was
    the last one to have the liberal BSD-3 clause license seems to be a good idea.
    
    Now does the full license as posted by Arman above need to be in each file
    or can it be like other ported code and have one copy in License.md. I would probably
    add something like "Full license text in LICENSE.md" to the SHA note above, just
    to tie the two together.

2) The starting class is, as Wojciech posted above, ` java.nio.file.ZipFileSystemProvider`.

   - by Java naming conventions, the implementing files would be in the `java.nio.file` directory.

   - following prior Scala Native art,  `ZipFileSystemProvider` would be in a new `java.nio/fs/zip/` directory.
     This is would be analogous to the existing `nio/fs/unix` and `nio/fs/windows` directories. Find one
     "out of place" implementation and one finds the others.

**Comment by LeeTibbert** [comment]
Implementation notes:

[ ] `user/setup.html` ReadTheDocs User Environment setup almost certainly needs to be
     edited to describe installing `zlib` on macOS. The easy (sic!) brew installation might
     introduce some macOS version limitations.  This setup (and, later, possibly more) should

**Comment by armanbilge** [comment]
> To that end, let me ask: Is anybody currently working on this issue? If yes, godspeed.
If no, I can take a run at it.

I am not, please go for it 🚀 

> A comment on the top of each zip file stating the commit SHA and that it was
the last one to have the liberal BSD-3 clause license seems to be a good idea.

👍 I agree. We can also include the `Copyright (c) 2009, 2010, Oracle and/or its affiliates` ...

> Now does the full license as posted by Arman above need to be in each file
or can it be like other ported code and have one copy in License.md

... in which case it seems reasonable to point to `License.md` for the full license.

> The starting class is, as Wojciech posted above, ` java.nio.file.ZipFileSystemProvider`.

Is it? In the JDK it is actually at `jdk.nio.zipfs.ZipFileSystemProvider`, should we be matching that?

**Comment by LeeTibbert** [comment]
If I understand correctly  the package name never gets exposed in an API. That is `javalib/src/main/scala/java/nio/file/spi/FileSystemProvider.scala#installedProviders` returns a list of `installedProviders`, which are anonymous 
subclasses on which one can call `newFileSystem()` or others.

So, the name of the subclass becomes a SN internal matter. Which guidon do we wish to follow?
What will make it easier for reviews and, especially, future maintainers?   What will make it easier
for someone familiar with the JDK to find the implementation file?

I could use `scala.scalanative.nio.fs.zipfs` or start a new, related convention
`scala.scalanative.jdk.nio.zipfs` with the `jdk.nio.zipfs` package name. If the latter,
I might leave a `ReadMe.txt` broken heart in `scala.scalanative.nio.fs` to help
future maintainers, like me, find the `jdk.nio.zipfs` package.


The existing providers.
`package scala.scalanative.nio.fs.unix`
`package scala.scalanative.nio.fs.windows`

**Comment by WojciechMazur** [comment]
The API of `jdk.nio.zipfs` actually exposes 3 public classes: `ByteArrayChannel`, `ZipFileSystemProvider` and `ZipInfo`. These 3 need to have a matching package name. Other stuff is internal, and can be put under different package. 
When it comes to directory in which it should be put in, then `scala/scalanative/nio.fs.*` convenction might be good to follow since it's clear to the editors of the code that this part is not part of the JDK public API without the need of checking the sources content. Either of the two layouts would work, and each of them makes sense to me.

**Comment by LeeTibbert** [comment]
Does anyone know the zlib situation and/or conventions for Windows off the top of their head?
Windows must do something with zip & jar files so there must be something, but probably not
zlib as such.

If not, I can crawl the Web after I get Linux & macOS working.

**Comment by WojciechMazur** [comment]
It is actually a zlib, it is used in our javalib for Zip{Input/Output}Stream, however the memory layout might be slightly different - one uses ushort/uint, other one uint/ulong for some parts of its structs (see `zlib.scala` gz_header)

**Comment by LeeTibbert** [comment]
@WojciechMazur At_Community,

The month anniversary of the base Issue is a good time for an update.

This request is bigger than a bread box.  I invested several weeks of effort an made some
progress before being pulled in different directions.    

##### Design decision required

###### TL; DR I believe zipfs need to be under a configuration option, probably LinktimeInto

A straight forward, sight unseen, implementation of zipfs adds a line of code to 
`javalib/src/main/scala/java/nio/file/spi/FileSystemProvider.scala`
```
object FileSystemProvider {
  def installedProviders: List[FileSystemProvider] = {
    val list = new LinkedList[FileSystemProvider]
    if (isWindows)
      list.add(new WindowsFileSystemProvider())
    else
      list.add(new UnixFileSystemProvider())                 
                                   
    list.add(new jdk.nio.zipfs.ZipFileSystemProvider()) // Support ZipFileSystem               
    list
  }

```

This has a short cascade of follow-on consequences:

1) the symbol `jdk.nio.zipfs.ZipFileSystemProvider()` is always used so it and
    everything it references is not subject to Dead Code Elimination.  This add
    a __significant__ amount of code to each add every resultant binary.  99.99%
    of those have less than zero interest in ZipFS.

2) Some of the code which can not be eliminated is the `java.util.zip` code.
    That code has a link time dependency on a zip lib being installed on the
    linking system. I have not checked the code to see if the link is static or
    dynamic.  The latter means a smaller binary, but also means that a 
    zip lib must be available on the runtime system.

    Scala native ReatTheDocs, Version 0.4.16-SNAPSHOT, describes:
    ```
    zlib 1.2.8 or newer (optional)

   These are only required if you use the corresponding feature.
   ```
   and
   ```
   If you use classes from the java.util.zip for compression zlib needs to be installed.
   ```

   A straight forward implementation of zipfs would make this description obsolete.


Please advise.  

I have other pressing work and this is a months long task, but I would like to keep the stone/discussion
moving, if only at a gopher tortoise pace.

**Comment by LeeTibbert** [comment]
JDK Zip file system support appears to be stable into the foreseeable future.
Any sketch/design of Zip file system support, meaning reading .jar files, should
be aware that the `.jar` format gained a sibling/successor of `.jrt` in Java 9, 
[java runtime format](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java.base-summary.html)

While `.jrt` support should not prevent timely `.jar` support, it does need to be factored into any 
mid-to-long range plan and tracked as such.

The major implication for the short term is that the framework for `.jar` support is quite likely to 
be used for `.jrt` support. Meaning that it needs to be reasonably clean itself and rationalized with
the existing ./javalib/src/main/scala/scala/scalanative/nio/fs/GenericFileSystemProvider.scala`.

**Comment by armanbilge** [comment]
Thanks for all your work and the detailed write-up. 

> TL; DR I believe zipfs need to be under a configuration option, probably LinktimeInto

Yes. Actually, and we should expand this to a separate issue/discussion, but I believe it is worthwhile to solve this more generally than just zipfs: the JDK uses "providers" to make various implementations available at runtime (file systems, sockets, crypto, etc.) and we should consider designing a linktime equivalent of this system.

As if Scala Native platform support isn't already complicated enough, it may get even more complicated as we start to support WASM etc. Having linktime-configurable providers could open the door for implementations of these APIs to exist _outside_ the Scala Native repository and be developed and maintained by the community instead of the core team.

**Comment by LeeTibbert** [comment]
@armanbilge Good perspective, thank you.

It would be nice if the whole whack of zipFs code were an artifact, compiled once a month or so, instead
of each and every application having to compile it only to have Dead Code Elimination discard it.  

In my University days, we started a term with a theoretical 'budget' for computer time.  Each (batch) run was
charged against that. The fear was always that one would run out before the end of term.  I think the
psychology of scarcity  was instilled in me by that (repeated trauma). To this day, I hate to see CPU cycles
and network transfer bytes go to waste.

**Comment by LeeTibbert** [comment]
@armanbilge 

To try to push the peg, in fits & starts.

I think there are at least two approaches that could be sketched out or roughly designed.

When designing, I always like to carry along at least two cases-under-implementation. That 
keeps from "overfitting" to the case at hand.  

I have always skipped over a "sockets" provider.  If I can create some spare cycles, I might look at
that.  Or else a "crypto" provide; crypto tends to be an all-consuming swamp.

Getting zipfs to be trustworthy is going to be, best case, the work of months. If there is some
urgency to getting native tools working, it may make sense to go with a simple LinktimeInfo
(say "Comment" or such, to avoid tying too closely to zipfs) at a first implementation and
keep pushing a larger/better solution.


1) Link time providers 

    In the zipfs case the supporting files could be in javalib.  I would have to trace out if
    an alternate implementation could be resolved from a quasi-3rd party artifact.

    The difficulty with having the "supporting" files in javalib is that they get compiled
    and then thrown away 99% of the time.  Yes, they could have a LinktimeInfo check so
    that only a minimal file got compiled but still.



2) A provider as a loader of a sharable library

     I believe that SN 0.5.0, and possibly higher version of 0.4.n can be linked as a sharable library (.so).

     The "Provider" could be a thin Scala layer which loaded an indicated shareable image, where the
     shareable image contained compiled & linked Scala code.   This should work on Linux, macOS, Windows,
     and FreeBSD.  It would take someone who knows Windows and has access to a Windows build system
     to implement on that OS.

     I would have to check:
     1) What are the dynamic garbage collection implications?  Especially if the main image is using one
          gc and the other something else?

     2) How to ensure that the multithreading status of each part is compatible.,

     3)  mumble Continuations mumble.  

     4)  packaging and accessing the shareable library needs to be addressed.

         I seem to remember that, at least on Linux, using a sharable library that is not in one of the "standard places"
         takes special handling.  Compiling in the list of "standard places" probably requires different strings on
         different operating systems.  There is some magic which allows specifying the location of the 
         sharable library when the main image is run.   LD_LIBRARY mumble and/or a command line flag.

         There is a good reason why it has been ages since I worked with this stuff.

    The disadvantage of this approach is that it cuts off folks who hate shareable libraries and use only
    static libraries.

**Comment by armanbilge** [comment]
> 1. Link time providers
>        In the zipfs case the supporting files could be in javalib.

Yes, I believe this is the correct approach. A shared library approach would require some kind of ABI.

> is that they get compiled
>        and then thrown away 99% of the time.  Yes, they could have a LinktimeInfo check so
>        that only a minimal file got compiled but still.

Perhaps I'm misunderstanding, but this seems fine. We compile it once to NIR when we release javalib. Then most user applications will completely elide it so it will come at no cost to them, and only applications in which it is linktime reachable will need to further compile it to native code.

**Comment by armanbilge** [comment]
> When designing, I always like to carry along at least two cases-under-implementation. That keeps from "overfitting" to the case at hand.
> 
> I have always skipped over a "sockets" provider. If I can create some spare cycles, I might look at that. Or else a "crypto" provide; crypto tends to be an all-consuming swamp.

How about looking at `CharsetProvider`? `Charset` is something we already offer in core, and is hopefully innocuous enough and not too swampy 😅 (I am certain I will eat my words)

https://docs.oracle.com/javase/8/docs/api/java/nio/charset/spi/CharsetProvider.html

**Comment by LeeTibbert** [comment]
@WojciechMazur 

It looks like recently merged PR #3574 unblocks this Issue.  You did a lot of hard work on that PR.

To co-ordinate:

1) Is zipfs still useful & needed?

2) Are you planning on working on it or should I resume my work?

    a) If you are planning on working on it, there is some some Channels based UDP networking work
        I have been trying to get to for a long while.

    b) If you would like me to take another run, I need  two weeks or so to clear my current SN promises.

###### Thoughts
1) I think the next area that needs to be worked out is how to handle `unit-tests` both for default `opt-out`
    and for explicit `opt-in` zipfs support.   The former appears to be no-problem.   The latter may need to be
    done in a `scripted` test (yuch!!) or a separate project (unit-tests-zipfs) so that the "closed world" includes
    zipfs.  I never quite trust  "save/restore" scenarios, especially when testing.

2) As mentioned somewhere earlier in this topic, I have an ugly private design study based on the code Arman
    described above.  This "bottom up" approach works but leads to great shagginess & inconsistency with
    current SN code.

    If I were to take another run, I think I would try a new private study which starts at the top with loading 
    a Service Provider which is based on the current File Service Provider and then works its way down trying 
    to stay close to existing practice & code. This is a "top down" approach.  I think it will be much easier to 
    get right and to maintain over time.   

Please advise. Thank you.

**Comment by WojciechMazur** [comment]
> Is zipfs still useful & needed?

In my opinion it would be rally usefull to have it. We can try to workaround it by unzipping the jar to local directory and working on it that way, but it might give us some unnecessary overhead. However, even if we'd do so, there still might be a use case for it outside the SN toolchain - e.g. as an utility for for tools compiled using SN. So I'd say it's still usefull and it would be great to have it. 

> Are you planning on working on it or should I resume my work? 

Currently, I'd like to focus on providing the JDBC compat layer for the ongoing effort to create a multi-platform sql library for the Scala toolkit and stabilisation of remaining flows in the current build so we can finally release 0.5.0. It would be great if you could work on that, but because we can think about workaround we shouldn't give it a high priority. Channels/UDP  related work might be more important now. Thank you for all your efforts! 

I think that ServiceProvider approach might work better for us. If the zipfs would not be available we would simply end up with a runtime exception similarly to the JVM - in theory zipfs could be opt-in as well, but itis enable by default becouse it might be used by many internals (classloaders, resources). In such case testing can be made easier becouse we would never directly call zipfs classes, but rather sue intermediate FileSystemProvider. At that point we should be able to simply add assumption before test to check if zipfs is available and installed.

**Comment by LeeTibbert** [comment]
@WojciechMazur   Thank you for the guidance & co-ordination.  

Given the discussion to day, I think I will focus on draining my work queue (irritations: stdlib.scala, FreeBSD, etc), 
and then sketching out a work/dream plan for java.net channels, UDP, & TCP.  

If I get an odd weekend day, I might do some private studies of getting a dummy zip FileServiceProvider
and a dummy unit-tests calling each other without bring in the entire world. I need to educate myself &
get with the program. From reading alone, I am still concerned with adding zipfs weight/overhead to 
"normal" applications.

I can put any or all of this on hold if there is something I can do to help prepare for an 0.5.0 release,
in the fullness of time. Now that would be cause for rejoicing.

Godspeed on " JDBC compat layer".

**Comment by LeeTibbert** [comment]
OK I have thrashed at service loading for a day and it is (well past) time to admit lack of cleverness and
ask for help.  This is supposed to be a "5 minute" question, I do not want to derail more important matters.

It looks like trying to reverse engineer `LinktimeIntrinsicCallsResolver.scala` would be pretty time consuming
for me.

Goal: To be able to load a very preliminary `ZipFileSystemProvider` by getting the  `.withServiceProviders()`
          declaration right.

Given:

      A dummy ZipFileSystemProvider in `javalib/src/main/scala/scala/scalanative/nio/fs/zipfs/ZipFileSystemProvider.scala`
      which declares:
```
package scala.scalanative.nio.fs.zipfs

class ZipFileSystemProvider extends GenericFileSystemProvider {
// and a few overrides, such as `getScheme`
}
```

Code in `java/nio/file/spi/FileSystemProvider.scala` 
```
import scala.scalanative.nio.fs.unix.UnixFileSystemProvider
import scala.scalanative.nio.fs.windows.WindowsFileSystemProvider
import scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider

object FileSystemProvider {
  def installedProviders: List[FileSystemProvider] = {
    val list = new LinkedList[FileSystemProvider]
    val defaultProvider =
      if (isWindows) new WindowsFileSystemProvider()
      else new UnixFileSystemProvider()

    list.add(defaultProvider)
    ServiceLoader.load(classOf[FileSystemProvider]).forEach(list.add(_))

    ServiceLoader.load(classOf[ZipFileSystemProvider]).forEach(list.add(_))

    list
  }
}

```

and code in `project/Build.scala`
```
           Map(
              "org.scalanative.testsuite.javalib.util.MyService" -> Seq(
                "org.scalanative.testsuite.javalib.util.MyServiceImpl1",
                "org.scalanative.testsuite.javalib.util.MyServiceImpl2"
              ),
              "scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider" ->	Seq(
                "scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider")
	    )
          )
```

It looks like I have a total lack of understanding about what the  `Build.scala` addition should be,
because the ZipFileSystemProvider is never loaded.  I think I have the second line correct but
probably have word salad for the first line.

On building I get: the ZipfsProvider in the left column and `not loaded`, so something is being recognized.

I also futzed around with trying to add the ZipFile system to the list of java.nio providers (in the hope
that the existing `ServiceLoader.load(classOf[FileSystemProvider]).forEach(list.add(_))` code would
pick it up.
 ```
         "java.nio.file.spi.FileSystemProvider" -> Seq(
                "scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider")

```

Sorry to display my total lack of understanding.

Note: The name that I would like to expose to the public is the `jdk.zipfs` package name used in the Java > 8 
           documentation.  That can be a future refinement, if possible.

**Comment by WojciechMazur** [comment]
Let's start with the outputs of toolchain. 

```

|-------------------------------------------------------------------------------------------------------------------------------|
| Service                                          | Service Provider                                      | Status             |
|-------------------------------------------------------------------------------------------------------------------------------|
| java.nio.charset.spi.CharsetProvider             | ---                                                   | NoProviders        |
|                                                  |                                                       |                    |
| org.scalanative.testsuite.javalib.util.MyService | NonExistingServiceImplementation                      | UnknownConfigEntry |
|                                                  | org.scalanative.testsuite.javalib.util.MyServiceImpl1 | Loaded             |
|                                                  | org.scalanative.testsuite.javalib.util.MyServiceImpl2 | Loaded             |
|                                                  | org.scalanative.testsuite.javalib.util.MyServiceImpl3 | Available          |
|-------------------------------------------------------------------------------------------------------------------------------|
```

For `java.nio.charset.spi.CharsetProvider` we have not found any implementation of this interface on the classpath, so it's status is `NoProviders`. In runtime we would always result in empty list of providers when running `ServiceLoader.load(classOf[spi.CharsetProvider)` 

`org.scalanative.testsuite.javalib.util.MyService` for this interface we've defined following entry in nativeConfig 
```scala
            Map(
              "org.scalanative.testsuite.javalib.util.MyService" -> Seq(
                "org.scalanative.testsuite.javalib.util.MyServiceImpl1",
                "org.scalanative.testsuite.javalib.util.MyServiceImpl2",
                "NonExistingServiceImplementation"
              )
```

We have explicitly enabled `MyServiceImpl1` and `MyServiceImpl2`. These would be available at runtime. We have not enabled `MyServiceImpl3` - it was found on classpath as a subclass of `MyService` and toolchain hints about available implementaiton. `MyServiceImpl3` would not be return from `ServiceLoader.load(classOf[MyService)`. We have also defined 1 implementation in config which does not exist on classpath `NonExistingServiceImplementation` might be a typo or might suggest missing dependency. 

---

From the users perspective we would always start with all providers disabled. Based information the toolchain, user would get info about list of referenced SPIs and a list of their implementations (providers) found on classpath. User would decide which of them should be enabled based on this information.

From the perspective of library author, we need to provide a list of defined service implementations in `META-INF/services`
To notify toolchain about ZipFS implementation we'd need to create a file `javalib/src/main/resources/META-INF/services/java.nio.file.spi.FileSystemProvider` and populate it with list of implementors of this provider, so in our context it would contain single entry `scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider`
We'd like to test it in our unit tests, so we'd enable this custom file system in NativeConfig of `unit-tests`: 
```scala
// _.withServiceProviders(Map(interface -> implementations))
_.withServiceProviders(Map("java.nio.file.spi.FileSystemProvider" -> Seq("scala.scalanative.nio.fs.zipfs.ZipFileSystemProvider"))
```
---

> Do I need to be putting something in a META-INF file somewhere

Yes, that's probably the part that was missing. Both JVM and Scala Native needs a hints about available implementaions of given services. Otherwise both of them would need to load every single symbol on class path and check if it a subclass of this service. That why implementors of services need to create a `META-INF/services/<service_interface>` to publish this meta info to JVM/toolchain

Note: 
I'm assuming that we'd have somewhere defined class
```
package scala.scalanative.nio.fs.zipfs
(final?) class ZipFileSystemProvider extends java.nio.file.spi.FileSystemProvider
```
We don't want any direct references to `ZipFileSystemProvider` in our code - any reference would for their classloading. We'd only use `ServiceLoader.load(classOf[FileSystemProvider])` to register it in list of available FSProviders, and delegate creation of their instances via `FileSystems.newFileSystem` based on the `schema` assigned to each of FSProviders

**Comment by LeeTibbert** [comment]
Wojciech,

Thank you for the timely & helpful information. I appreciate you time anytime, but especially on your
weekend.

* The clear description of the need for `META-INF/services/<service_interface>` is the magic I needed.
   The `// _.withServiceProviders(Map(interface -> implementations))` description & concrete example
    help understanding.

* With this magic, I now have a sandbox `ZipFileServiceProvider` loading (in unit-tests) and executing
    tests which opens a zipfile and then reads and correctly displays the contents of entry, say
    `/META-INF/MANIFEST.MF`.  
    
    I am currently debugging some Java compatibility issues around "helpful" transformations
    done to the name evident to the file (adding & subtracting leading & trailing slashes).  Grist
    for the mill.

    When I have figured that out, hear the hopeful note there?, I will probably test creating a
    zip file, writing to it, and then reading the contents back.  After that it is on to testing
    the other methods of any self respecting `FileSystemProvider`.

    All that ought to keep me busy for awhile.

* With your consent, I will probably fold some of the description above into the existing 
   sections in javalib.rst and the mumble-0.5.0-SNAPSHOT.md.  
   If it helped me get up and running, it may help others.
   
* The `ServiceLoad` implementation is pretty elegant, once one gets it working.
   I noticed that you changed `java.nio.charset.Charset` to use `ServiceLoader`.
   I wonder what the Charset for Unicode emojis is?  (joke)

**Comment by LeeTibbert** [comment]
Progress update:

1) I have private sandbox code which loads a jdk.zipfs ServiceProvider. That provider can add
    directory entries and files in those directories.  These entries persist after the zipfs is closed
    and can be re-read by that code.

    Unfortunately, the file written can not be read by "unzip -1" (minus ell). Oops!

2) I have some private sandbox code which can read in a .zip file and write out a true copy.  That
    copy can be read by "unzip -l" and its relatives.

The next step is to modify the code in #2 above to be able to Create-Read-Update-Delete-List and write
out a .zip which can be read by standard zip utilities.

There is a huge amount of impenetrable  code involved.  My direction at this point is to use the existing
`java.util.zip` code, where possible and fix it as necessary.  My thought/hope is that this will lead to 
one impenetrable heap of code rather than two, each with its own bugs.

**Comment by LeeTibbert** [comment]
It is going on six months since I have been able to work on this Issue and I do not
see a chance to work on it in the next year or more.

It has become clear to me that I should accept the ding to my reputation and withdraw
from this task and not block or complicate work by others.

<hr>
Like the I leave behind a pile of weathered, broken, bones.

I think the approach of implementing `ZipFileSystem` in terms of
the existing and or fixed `java.util.zip` methods and being
relatively consistent with the existing `FileSystem` implemention
is going to lead to both more correct and, especially, more
maintainable code.

Zip has three or four decades of quirks, version, historical, and OS
versions.  Finding and fixing these in one place and using them in
at least two is more likely to be successful than having two sets
of bugs.

Part of the problem is that the current `java.util.zip` code has
_many_ bugs.

I tried Arman's suggestion:
```
The last commit with the liberally-licensed demo codes is here:
https://github.com/openjdk/jdk/tree/492411d1c6d8b34f3df00f152c9e4ad0ad5f97e3/jdk/src/share/demo/nio/zipfs
```

 I found that it provided a good demo 'proof-of-concept' but I could not get it to pre-alpha
or even sandbox quality and which would be believably maintainable. Perhaps a lack
of skill on my part.

IIRC, I got a sandbox "call into java.util.zip" implementation somewhat working, but found
I was spending my time on 'java.util.zip' bugs. I created PRs for each as I encountered them.

 Now I can close all the Firefox tabs referencing Zipfs I have up and which guilt trip me
everytime I restart Firefox. Some people  have a strange idea of relief.

**Comment by He-Pin** [comment]
@WojciechMazur Do you know if there are any easy way to install zlib on windows?
refs: https://github.com/databricks/sjsonnet/issues/391
